components/tidb_query/src/rpn_expr/impl_string.rs

```diff
@@ -111,6 +112,22 @@ pub fn locate_binary_2_args(substr: &Option<Bytes>, s: &Option<Bytes>) -> Result
         .or(Some(0)))
 }
 
+#[rpn_fn]
+#[inline]
+pub fn un_hex(arg: &Option<Bytes>) -> Result<Option<Bytes>> {
+    Ok(arg
+        .as_ref()
+        .map(|bytes| {
+            let mut bytes = bytes.clone();
```

Comment:
How can I avoid `clone()`?

Reply:
The roadmap is:

1. If `bytes` contains odd characters, unhex the first byte specially.

2. Unhex the rest bytes.

However, it's worth taking note of that:

1. Insert in front of a vector is not cheaper than clone it. As the [implementation](https://doc.rust-lang.org/src/alloc/vec.rs.html#951-974) of `insert` is to copy the rest vector.

2. `append` a long vector after a short one is also as expensive as clone the long one. [Implementation](https://doc.rust-lang.org/src/alloc/vec.rs.html#1159-1164)

Reply:
So in this implementation vector is copied three times: `bytes.clone()`, `bytes.insert(0, b'0')`, `Vec::from_hex(hex)`. The beginning two `clone` can be simply combined into one. 

However, reducing the "copy" from two to one is more complicate. You should append to a vector while calculating unhex rather than generate a unhexed vector and append it to the existing short vector later.

Good news is that the newer version of `hex` has function [`decode_to_slice`](https://docs.rs/hex/0.4.0/hex/fn.decode_to_slice.html). It can help you avoid this "copy". (Remember to preallocate enough space in vector)

Reply:
Thank you for your explanation and help. I'll change it again